### PR TITLE
Refs #20483 --  Lifted available_apps limitation in regard to unavailable models references.

### PR DESCRIPTION
**Pull request by charettes**

### CODE REVIEW

**Comment by charettes** [comment]
buildbot, test on oracle.

File: tests/fixtures/tests.py [hunk 1]
```diff
@@ -655,7 +655,6 @@ class NonExistentFixtureTests(TestCase):
     """
     Custom class to limit fixture dirs.
     """
-    available_apps = ['django.contrib.auth', 'django.contrib.contenttypes']
```
Comment (charettes) [inline tests/fixtures/tests.py]: Defining `available_apps` on a `TestCase` has not effect.

**Comment by charettes** [comment]
buildbot, test on oracle.

File: tests/model_inheritance/tests.py [hunk 2]
```diff
@@ -399,48 +398,37 @@ def test_exclude_inherited_on_null(self):
...
+class InheritanceSameModelNameTests(SimpleTestCase):
+    def test_abstract_fk_related_name(self):
+        class Referenced(models.Model):
+            class Meta:
+                app_label = 'model_inheritance'
 
-    available_apps = ['model_inheritance']
+        class AbstractReferent(models.Model):
+            reference = models.ForeignKey(
+                Referenced, models.CASCADE, related_name='%(app_label)s_%(class)s_references'
```
Comment (timgraham) [inline tests/model_inheritance/tests.py]: I'd use a local variable for '%(app_label)s_%(class)s_references' so we don't risk typoing it in the assertion.

File: django/db/backends/oracle/operations.py [hunk 3]
```diff
@@ -310,17 +312,88 @@ def savepoint_create_sql(self, sid):
...
-            # FK referential trouble, whereas DELETE FROM table works.
-            sql = ['%s %s %s;' % (
-                style.SQL_KEYWORD('DELETE'),
-                style.SQL_KEYWORD('FROM'),
-                style.SQL_FIELD(self.quote_name(table))
-            ) for table in tables]
+            constraints = []
+            truncated_tables = {table.upper() for table in tables}
+            if allow_cascade:
+                # We can't use TRUNCATE CASCADE as it only works with ON DELETE
```
Comment (timgraham) [inline django/db/backends/oracle/operations.py]: Write comments without "we" p lease.

**Comment by charettes** [comment]
@shaib, did you find time to work on a recursive query to retrieve all dependent tables and constraints from an initial set of tables?

@timgraham, any objection to merge 957f4cabc1af02b2de285eda882f1421a4c22f08 separately now that I've assigned the `related_name` to a reused variable?

**Comment by shaib** [comment]
On Tuesday 31 May 2016 21:50:51 Simon Charette wrote:
> @shaib, did you find time to work on a recursive query to retrieve all
> dependent tables and constraints from an initial set of tables?
> 
Sorry, no. Not yet.

**Comment by charettes** [comment]
@felixxm do you think you could be able to come up with recursive query to retrieve all dependent tables and constraints from an initial set of tables on Oracle? Speeding up the `TRUNCATE CASCADE` implementation emulation is the only thing left to merge this PR.

**Comment by felixxm** [comment]
This query should return all dependent tables and constraints for specific table:
 
```python
cursor.execute("""
    SELECT
        user_tables.table_name, rcons.constraint_name
    FROM
        user_tables
    JOIN
        user_constraints cons 
        ON (user_tables.table_name = cons.table_name AND cons.constraint_type = ANY('P', 'U'))
    LEFT JOIN
        user_constraints rcons 
        ON (user_tables.table_name = rcons.table_name AND rcons.constraint_type = 'R')
    START WITH user_tables.table_name = UPPER(%s)
    CONNECT BY NOCYCLE PRIOR cons.constraint_name = rcons.r_constraint_name
    GROUP BY
        user_tables.table_name, rcons.constraint_name
    HAVING user_tables.table_name != UPPER(%s);
""", [table_name, table_name])
```
But we still need to keep appropriate order.
I'm going to prepare all change tomorrow.

**Comment by charettes** [comment]
Awesome @felixxm, thank you so much!

**Comment by felixxm** [comment]
I reorganized and simplified this PR. Please take a look at https://github.com/felixxm/django/commit/7477165acf67ad96b0665d63fe3feb32482c8e57.

**Comment by charettes** [comment]
Your changes are looking good!

Do you think it could be possible/desirable to make `_foreign_key_constraints()` accept a list of table names and use `table_name IN tables_names` to perform a single query instead?

**Comment by felixxm** [comment]
I think that won't work correct. We can reach the same constraint and foreign table few times (from different source tables) it can change `DML` ordering which is very important. That's why I build `sql` lists separately for each source table (see line [359](https://github.com/felixxm/django/commit/7477165acf67ad96b0665d63fe3feb32482c8e57#diff-ea23a0501b6f65fb447da98953a35dafR359)). This implementation works quite fast.

**Comment by charettes** [comment]
> This implementation works quite fast.

Awesome, I'll merge your changes in this PR and give you attribution later today!

**Comment by charettes** [comment]
Buildbot, test on oracle.

File: django/db/backends/oracle/operations.py [hunk 4]
```diff
@@ -315,17 +315,71 @@ def savepoint_create_sql(self, sid):
...
+            truncated_tables = {table.upper() for table in tables}
+            constraints = set()
+            if allow_cascade:
+                # Oracle's TRUNCATE CASCADE only works with ON DELETE CASACADE
+                # foreign keys which Django doesn't define. Emulate the
+                # PostgreSQL behavior which truncates all dependent tables by
+                # manually retrieving all foreign key constraints and resolving
+                # dependencies.
+                for table in tables:
+                    for foreign_table, constraint in self._foreign_key_constraints(table):
```
Comment (felixxm) [inline django/db/backends/oracle/operations.py]: IMO we need to split constraints for each source table (see [[1]](https://github.com/felixxm/django/commit/7477165acf67ad96b0665d63fe3feb32482c8e57#diff-ea23a0501b6f65fb447da98953a35dafR354) and [[2]](https://github.com/felixxm/django/commit/7477165acf67ad96b0665d63fe3feb32482c8e57#diff-ea23a0501b6f65fb447da98953a35dafR359)). `backends.tests.LongNameTest.test_sequence_name_length_limits_flush` passed with this approach.

Comment (charettes) [inline django/db/backends/oracle/operations.py]: I don't think it's necessary. This approach was truncating and switching constraints on/off multiple times if the table appeared in the dependency cycle more than once. The `backends.tests.LongNameTest.test_sequence_name_length_limits_flush` test is failing when calling sqlflush without `allow_cascade=True`.

Comment (felixxm) [inline django/db/backends/oracle/operations.py]: You're right I didn't notice that. As far as I'm concerned the last thing we missed is to add query for the first level constraints i.e.:
```python
cursor.execute("""
    SELECT
        cons.table_name, cons.constraint_name, 1
    FROM
        user_constraints cons
    WHERE
        cons.constraint_type = 'R'
        AND cons.table_name = UPPER(%s)
""", [table_name])
```
(see https://github.com/felixxm/django/commit/722863ac293072aa93c72f5ff08e6b699ebe4deb).

**Comment by charettes** [comment]
@felixxm I had similar code ready to merge but I was struggling with the second query construction, thanks for that!

Ideally `_foreign_key_constraints` would accept a list of table names and perform a single query to retrieve all the constraints (using `table_name IN ...`) but as this approach performs reasonably well I think the current behavior is acceptable.

**Comment by charettes** [comment]
Buildbot, test on oracle.

File: tests/multiple_database/tests.py [hunk 5]
```diff
@@ -1794,11 +1794,6 @@ def allow_migrate(self, db, app_label, **hints):
 
 class MigrateTestCase(TestCase):
 
-    available_apps = [
-        'multiple_database',
-        'django.contrib.auth',
-        'django.contrib.contenttypes'
-    ]
```
Comment (charettes) [inline tests/multiple_database/tests.py]: Defining `available_apps` on a `TestCase` has no benefit.

File: tests/swappable_models/tests.py [hunk 6]
```diff
@@ -11,12 +11,6 @@
 
 class SwappableModelTests(TestCase):
 
-    available_apps = [
-        'swappable_models',
-        'django.contrib.auth',
-        'django.contrib.contenttypes',
-    ]
-
```
Comment (charettes) [inline tests/swappable_models/tests.py]: Defining `available_apps` on a `TestCase` has no benefit.

**Comment by charettes** [comment]
@timgraham this should be good for a final review if you're not too swamped tomorrow. I don't mind if this doesn't make it in the 1.11 cut, this is really just a cleanup.

**Pull request merged**